packages/tools/devtools/devtools-core/src/data-visualization/SharedTreeVisualizer.ts

```diff
@@ -0,0 +1,250 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import type {
+	FieldMapObject,
+	JsonableTree,
+	SharedTreeContentSnapshot,
+	TreeNodeStoredSchema,
+	TreeTypeSet,
+} from "@fluidframework/tree/internal";
+import {
+	EmptyKey,
+	LeafNodeStoredSchema,
+	MapNodeStoredSchema,
+	ObjectNodeStoredSchema,
+} from "@fluidframework/tree/internal";
+import type { SharedTreeLeafNode, VisualSharedTreeNode } from "./VisualSharedTreeTypes.js";
+import { SharedTreeSchemaType, VisualSharedTreeNodeKind } from "./VisualSharedTreeTypes.js";
+import { type VisualChildNode, VisualNodeKind, type VisualValueNode } from "./VisualTree.js";
+
+/**
+ * Converts the output of {@link sharedTreeVisualizer} to {@link VisualChildNode} type containing `schema` and `children` fields.
+ */
+export function toVisualTree(tree: VisualSharedTreeNode): VisualChildNode {
+	if (tree.kind === VisualSharedTreeNodeKind.LeafNode) {
+		const result: VisualValueNode = {
+			value: tree.value,
+			nodeKind: VisualNodeKind.ValueNode,
+			tooltipContents: tree.schema.name,
+		};
+		return result;
+	} else {
+		const children: Record<string, VisualChildNode> = {};
+
+		for (const [key, value] of Object.entries(tree.fields)) {
+			const child = toVisualTree(value);
+			children[key] = child;
+		}
+
+		return {
+			children,
+			nodeKind: VisualNodeKind.TreeNode,
+			tooltipContents: tree.schema.allowedTypes,
+		};
+	}
+}
+
+/**
+ * Concatenrate allowed types for `ObjectNodeStoredSchema` and `MapNodeStoredSchema`.
+ */
+function concatenateAllowedTypes(fieldKey: string, fieldTypes: TreeTypeSet | undefined): string {
+	let fieldAllowedType = fieldKey === EmptyKey ? "" : `${fieldKey} : `;
+
+	if (fieldTypes === undefined) {
+		fieldAllowedType += "any";
+	} else {
+		for (const type of fieldTypes) {
```

Comment:
You should probably be able to replace this loop with `[...fieldTypes].join(" | ")` - then you wouldn't need to do any trimming afterward or anything.